### DISCUSSION TITLE
getConnectionPIPs not working for long wires

### DISCUSSION THREAD

**Discussion by sma-asghar** [Xilinx/RapidWright#802] [Q&A] [answered]
Hello Chris,

I am trying to find the INT tiles a net passes through. To filter the PIPs used by a particular sinkPin, I am using getConnectionPIPs().
However, it seems like the method doesn't work for long wires as they use bi-directional PIPs. 

I found the code for getConnectionPIPs() from an earlier discussion in the Google groups 'Split nets with multiple sink pins'. The problem seems to arise in the creation of a map from end nodes to PIP. For PIPs connected to long wires, the endNode of the preceding PIP is not the startNode of the following PIP.

Could you please suggest an alternative approach to get around this problem?

Thanks and regards,
Ali

**Comment by clavin-xlnx** [comment]
Hi Ali,

It seems like `DesignTools.getConnectionPIPs()` should do what you are describing.  If there is a bug in the method, we'd certainly like to fix it.  Could you share an example of how it is failing?  If you can't share a DCP, we have a new class (`CodeGenerator`, see #784) that will take a DCP/design and net(s) and will print out RapidWright code to reproduce just that net or scenario.  

Thanks,

Chris

**Reply by sma-asghar** [reply]
Thanks for your reply Chris. Could you please give an example of how to use the CodeGenerator? I couldn't find it on the API reference page.

Thanks,
Ali

**Reply by clavin-xlnx** [reply]
Sure, I took a look at the code and I actually had to add a bit more functionality for this case (it wasn't adding site pins).  So, if you aren't able to share a DCP, you can use this approach to provide a way to create just the net that might be causing the issue:

My changes haven't been merged yet, but they can be checked out by switching to the branch `add_site_pins`:
```
# From your RapidWright directory
git checkout add_site_pins
./gradlew compileJava
```
Here is an example program that extracts just a single net from a DCP to generate a RapidWright code example:
```java
package com.xilinx.rapidwright.examples;

import java.util.Collections;

import com.xilinx.rapidwright.design.Design;
import com.xilinx.rapidwright.tests.CodePerfTracker;
import com.xilinx.rapidwright.util.CodeGenerator;

public class ExampleCodeGenerator {

    public static void main(String[] args) {
        Design d = Design.readCheckpoint("test/RapidWrightDCP/microblazeAndILA_3pblocks.dcp", CodePerfTracker.SILENT);

        String code = CodeGenerator.testNetGenerator(d, Collections.singletonList(
                "base_mb_i/axi_gpio_0/U0/gpio_core_1/Not_Dual.gpio_Data_Out_reg[4]_lopt_replica_1"),
                Collections.emptyList());
        System.out.println(code);
    }
}
```
Running this program will print out the following RapidWright code snippet:
```java
Design design = new Design("base_mb_wrapper", "xcku035-ffva1156-2-e");
Device device = design.getDevice();

Net net = com.xilinx.rapidwright.util.CodeGenerator.createTestNet(design, "net", new String[]{
        "INT_X30Y76/INT.LOGIC_OUTS_E27->>INT_NODE_SINGLE_DOUBLE_4_INT_OUT",
        "INT_X30Y76/INT.INT_NODE_SINGLE_DOUBLE_4_INT_OUT->>NN2_E_BEG6",
        "INT_X30Y78/INT.NN2_E_END6->>INT_NODE_QUAD_LONG_5_INT_OUT",
        "INT_X30Y78/INT.INT_NODE_QUAD_LONG_5_INT_OUT->>NN16_BEG3",
        "INT_X30Y95/INT.NN16_BLN_3_FTS->>INT_NODE_QUAD_LONG_0_INT_OUT",
        "INT_X30Y95/INT.INT_NODE_QUAD_LONG_0_INT_OUT->>NN5_BEG0",
        "INT_X30Y100/INT.NN5_END0->>INT_NODE_QUAD_LONG_8_INT_OUT",
        "INT_X30Y100/INT.INT_NODE_QUAD_LONG_8_INT_OUT->>NN16_BEG0",
        "INT_X30Y116/INT.NN16_END0->>INT_NODE_QUAD_LONG_24_INT_OUT",
        "INT_X30Y116/INT.INT_NODE_QUAD_LONG_24_INT_OUT->>EE4_BEG2",
        "INT_X32Y116/INT.EE4_END2->>INT_NODE_GLOBAL_14_OUT1",
        "INT_X32Y116/INT.INT_NODE_GLOBAL_14_OUT1->>INT_NODE_IMUX_40_INT_OUT",
        "INT_X32Y116/INT.INT_NODE_IMUX_40_INT_OUT->>IMUX_W7",
        "INT_INT_INTERFACE_XIPHY_FT_X32Y116/INT_INT_INTERFACE_XIPHY_FT.IMUX7->>IMUXOUT7",
        "XIPHY_L_X32Y60/XIPHY_L.XIPHY_BITSLICE_TILE_407_TX_D0->>XIPHY_BITSLICE_TILE_407_TX_Q",
        "HRIO_L_X31Y90/HRIO_L.HRIO_IOB_44_OP->HRIO_IOB_44_OP_PIN"
});
SiteInst si1 = design.createSiteInst(device.getSite("SLICE_X47Y76"));
net.createPin("GQ", si1);
SiteInst si2 = design.createSiteInst(device.getSite("IOB_X1Y100"));
net.createPin("OP", si2);
```
You can then use this generated code to create an example test case for the issue you've described by copying and pasting it into a new Java class:
```
package com.xilinx.rapidwright.examples;

import com.xilinx.rapidwright.design.Design;
import com.xilinx.rapidwright.design.DesignTools;
import com.xilinx.rapidwright.design.Net;
import com.xilinx.rapidwright.design.SiteInst;
import com.xilinx.rapidwright.device.Device;

public class ExampleTest {

    public static void main(String[] args) {
        // -- Copied from ExampleCodeGenerator output
        Design design = new Design("base_mb_wrapper", "xcku035-ffva1156-2-e");
        Device device = design.getDevice();

        Net net = com.xilinx.rapidwright.util.CodeGenerator.createTestNet(design, "net", new String[] {
                "INT_X30Y76/INT.LOGIC_OUTS_E27->>INT_NODE_SINGLE_DOUBLE_4_INT_OUT",
                "INT_X30Y76/INT.INT_NODE_SINGLE_DOUBLE_4_INT_OUT->>NN2_E_BEG6",
                "INT_X30Y78/INT.NN2_E_END6->>INT_NODE_QUAD_LONG_5_INT_OUT",
                "INT_X30Y78/INT.INT_NODE_QUAD_LONG_5_INT_OUT->>NN16_BEG3",
                "INT_X30Y95/INT.NN16_BLN_3_FTS->>INT_NODE_QUAD_LONG_0_INT_OUT",
                "INT_X30Y95/INT.INT_NODE_QUAD_LONG_0_INT_OUT->>NN5_BEG0",
                "INT_X30Y100/INT.NN5_END0->>INT_NODE_QUAD_LONG_8_INT_OUT",
                "INT_X30Y100/INT.INT_NODE_QUAD_LONG_8_INT_OUT->>NN16_BEG0",
                "INT_X30Y116/INT.NN16_END0->>INT_NODE_QUAD_LONG_24_INT_OUT",
                "INT_X30Y116/INT.INT_NODE_QUAD_LONG_24_INT_OUT->>EE4_BEG2",
                "INT_X32Y116/INT.EE4_END2->>INT_NODE_GLOBAL_14_OUT1",
                "INT_X32Y116/INT.INT_NODE_GLOBAL_14_OUT1->>INT_NODE_IMUX_40_INT_OUT",
                "INT_X32Y116/INT.INT_NODE_IMUX_40_INT_OUT->>IMUX_W7",
                "INT_INT_INTERFACE_XIPHY_FT_X32Y116/INT_INT_INTERFACE_XIPHY_FT.IMUX7->>IMUXOUT7",
                "XIPHY_L_X32Y60/XIPHY_L.XIPHY_BITSLICE_TILE_407_TX_D0->>XIPHY_BITSLICE_TILE_407_TX_Q",
                "HRIO_L_X31Y90/HRIO_L.HRIO_IOB_44_OP->HRIO_IOB_44_OP_PIN"
        });
        SiteInst si1 = design.createSiteInst(device.getSite("SLICE_X47Y76"));
        net.createPin("GQ", si1);
        SiteInst si2 = design.createSiteInst(device.getSite("IOB_X1Y100"));
        net.createPin("OP", si2);
        // -- end

        System.out.println(DesignTools.getConnectionPIPs(si2.getSitePinInst("OP")));
    }
}
```

**Reply by sma-asghar** [reply]
Thanks for the detailed response Chris. Following the steps, I have managed to create an ExampleTest class. 
Running the code with this class generates the following error:
Exception in thread "main" java.lang.RuntimeException: ERROR: Failed to create site instance for site SLICE_X29Y174
	at com.xilinx.rapidwright.design.Design.createSiteInst(Unknown Source)

Luckily, the getConnectionPIPs works with these sitePins, therefore, I commented out the instantiation of all other siteInsts while keeping the problematic pin "C3" at siteInst  "si2". The code worked, reproducing the error I have seen with the dcp.

Please see the file RW_code_test

Thanks,
Ali
[RW_code_test.txt](https://github.com/Xilinx/RapidWright/files/12497410/RW_code_test.txt)

**Reply by clavin-xlnx** [reply, marked as answer]
Thanks Ali, I've created a PR (#809 ) that should hopefully resolve the issue for you.
